File: hoffman-agents-node/src/combination/operator.js
const crypto = require('crypto');
const { ExperienceTrie } = require('../core/experience-trie');
const { MetaTrie } = require('../core/meta-trie');
const { SelfTokenState } = require('../core/self-token');
const { TraceBuffer, TraceEvent } = require('../core/trace-buffer');
const { ExperienceLexicon } = require('../core/experience-lexicon');
const { ConsciousAgent } = require('../agent/conscious-agent');
const { ExperienceSpace } = require('../agent/experience-space');

function combine(...agents) {
  if (agents.length === 0) return trivialAgent();
  if (agents.length === 1) return agents[0];

  const hasWeights = agents.length === 2 && agents[1] && typeof agents[1] === 'object' && !(agents[1] instanceof ConsciousAgent) && 'weights' in agents[1];
  
  if (hasWeights) {
    const [agent1, opts] = agents;
    const agent2 = opts.other;
    const w1 = opts.weights[0];
    const w2 = opts.weights[1];
    return _weightedCombine(agent1, agent2 || opts.other || trivialAgent(), w1, w2);
  }

  if (agents.length === 2) {
    const [agent1, agent2] = agents;
    if (agent1.agentId === 'CA_0') return agent2;
    if (agent2.agentId === 'CA_0') return agent1;

    const combinedId = _hashAgentIds(agent1.agentId, agent2.agentId);
    const mergedTrie = agent1.experience.trie.merge(agent2.experience.trie);

    const leafIds1 = agent1.leafConstituentIds.size > 0 ? agent1.leafConstituentIds : new Set([agent1.agentId]);
    const leafIds2 = agent2.leafConstituentIds.size > 0 ? agent2.leafConstituentIds : new Set([agent2.agentId]);
    const leafConstituentIds = new Set([...leafIds1, ...leafIds2]);
    const constituentIds = new Set([agent1.agentId, agent2.agentId]);

    const jointMt = _buildJointMetaTrie(agent1.experience.metaTrie, agent2.experience.metaTrie);
    const combinedSelf = _combineAttractors(agent1.experience.selfToken, agent2.experience.selfToken);
    const mergedLexicon = _mergeLexicons(agent1.experience.lexicon, agent2.experience.lexicon);

    let initState = null;
    const traceBuf = new TraceBuffer(50);

    // Transfer short-term memory from the preferred parent (mt1) to preserve continuity
    const sourceBuf = agent1.experience.traceBuffer;
    const recentEvents = sourceBuf.getRecent(10);
    for (const e of recentEvents) {
      traceBuf.append(e);
    }
    if (recentEvents.length > 0) {
      initState = recentEvents[recentEvents.length - 1].toState;
    } else if (agent2.experience.lastWorldStateId !== null) {
      initState = agent2.experience.lastWorldStateId;
    } else {
      initState = _canonicalHash(-1, -1);
      traceBuf.append(new TraceEvent(-1, initState, 0, -1, false, 0.5));
    }

    const exp = new ExperienceSpace({
      trie: mergedTrie,
      metaTrie: jointMt,
      selfToken: combinedSelf,
      lexicon: mergedLexicon,
      traceBuffer: traceBuf,
      lastWorldStateId: initState,
    });

    return new ConsciousAgent({
      agentId: combinedId,
      experience: exp,
      generation: Math.max(agent1.generation, agent2.generation),
      constituentIds,
      leafConstituentIds,
      cycleLevel: Math.max(agent1.cycleLevel, agent2.cycleLevel) + 1,
    });
  }

  const mid = Math.floor(agents.length / 2);
  return combine(combine(...agents.slice(0, mid)), combine(...agents.slice(mid)));
}

function _weightedCombine(agent1, agent2, w1, w2) {
  const base = combine(agent1, agent2);
  const total = w1 + w2;
  if (total === 0) return base;
  const normW1 = w1 / total;
  const normW2 = w2 / total;
  const weightedStable = agent1.pStable * normW1 + agent2.pStable * normW2;
  const weightedLexicon = agent1.pLexicon * normW1 + agent2.pLexicon * normW2;
  const weightedExplore = agent1.pExplore * normW1 + agent2.pExplore * normW2;
  base.pStable = weightedStable;
  base.pLexicon = weightedLexicon;
  base.pExplore = weightedExplore;
  return base;
}

function trivialAgent() {
  return new ConsciousAgent({ agentId: 'CA_0' });
}

function experienceSpaceDistance(exp1, exp2) {
  const p1 = new Set(exp1.trie.getAllPaths(0).map(p => p.join(',')));
  const p2 = new Set(exp2.trie.getAllPaths(0).map(p => p.join(',')));
  const union = new Set([...p1, ...p2]);
  if (union.size === 0) return 0;
  const intersection = new Set([...p1].filter(x => p2.has(x)));
  const trieDist = 1 - intersection.size / union.size;

  const l1 = new Set(exp1.lexicon._entries.keys());
  const l2 = new Set(exp2.lexicon._entries.keys());
  const unionL = new Set([...l1, ...l2]);
  const lexiconDist = unionL.size === 0 ? 0 : 1 - new Set([...l1].filter(x => l2.has(x))).size / unionL.size;

  return trieDist * 0.6 + lexiconDist * 0.4;
}

function _hashAgentIds(id1, id2) {
  return `CA_${crypto.createHash('sha256').update(`${id1}->${id2}`).digest('hex').slice(0, 12)}`;
}

function _canonicalHash(s1, s2) {
  const sorted = [s1, s2].sort((a, b) => a - b);
  const h = crypto.createHash('sha256').update(`ws:${sorted[0]}:${sorted[1]}`).digest();
  return h.readUInt32BE(0);
}

function _buildJointMetaTrie(mt1, mt2) {
  const joint = new MetaTrie(
    Math.max(mt1._snapshotWindow, mt2._snapshotWindow),
    Math.max(mt1.trie.maxDepth, mt2.trie.maxDepth)
  );

  for (const [mid, snap] of mt1._registry) {
    joint._registry.set(mid | 0x10000000, snap);
  }
  for (const [mid, snap] of mt2._registry) {
    joint._registry.set(mid | 0x20000000, snap);
  }

  const _transferTrie = (sourceTrie, mask) => {
    const paths = sourceTrie.getAllPaths(1);
    for (const path of paths) {
      if (path.length < 2) continue;
      const shifted = path.map(id => id | mask);
      joint._trie.insert(shifted);
      const sourceNode = sourceTrie.lookup(path);
      const jointNode = joint._trie.lookup(shifted);
      if (sourceNode && jointNode) {
        jointNode.visitCount = sourceNode.visitCount;
        jointNode.predictionErrors = [...sourceNode.predictionErrors];
        jointNode.meanPredictionError = sourceNode.meanPredictionError;
      }
    }
  };

  _transferTrie(mt1.trie, 0x10000000);
  _transferTrie(mt2.trie, 0x20000000);

  if (mt1.lastMetaState !== null) {
    joint._lastMetaState = mt1.lastMetaState | 0x10000000;
  } else if (mt2.lastMetaState !== null) {
    joint._lastMetaState = mt2.lastMetaState | 0x20000000;
  }

  for (const [mid, counts] of mt1._tokenRegistry) {
    joint._tokenRegistry.set(mid | 0x10000000, new Map(counts));
  }
  for (const [mid, counts] of mt2._tokenRegistry) {
    joint._tokenRegistry.set(mid | 0x20000000, new Map(counts));
  }

  return joint;
}

function _combineAttractors(st1, st2) {
  return new SelfTokenState({
    token: st1.token,
    lockThreshold: Math.min(st1.lockThreshold, st2.lockThreshold),
    lockConsecutiveRequired: Math.max(st1.lockConsecutiveRequired, st2.lockConsecutiveRequired),
    stationaryProb: (st1.stationaryProb + st2.stationaryProb) / 2,
    locked: st1.locked && st2.locked,
    referentMetaStateId: st1.locked && st2.locked ? (st1.referentMetaStateId | 0x10000000) : null,
    lockGeneration: st1.locked && st2.locked ? Math.max(st1.lockGeneration || 0, st2.lockGeneration || 0) : null,
  });
}

function _mergeLexicons(lex1, lex2) {
  const merged = new ExperienceLexicon(
    Math.max(lex1._embeddingDim, lex2._embeddingDim),
    Math.min(lex1._associationThreshold, lex2._associationThreshold)
  );

  for (const entry of [...lex1._entries.values(), ...lex2._entries.values()]) {
    merged.bind(entry.label, entry.traceSignature, {
      predictionErrorPeak: entry.predictionErrorPeak,
      source: entry.labelingSource,
      generation: entry.generationBound,
      step: entry.stepBound,
      outputToken: entry.outputToken,
    });
    const e = merged._entries.get(entry.label);
    if (e) {
      e.integrationDepth = entry.integrationDepth;
      e.encounterCount = entry.encounterCount;
    }
  }
  return merged;
}

function _splitMetaTrie(jointMt, mask) {
  const stripped = new MetaTrie(jointMt._snapshotWindow, jointMt.trie.maxDepth);

  for (const [mid, snap] of jointMt._registry) {
    if ((mid & mask) !== 0) {
      stripped._registry.set(mid & ~mask, snap);
    }
  }

  const paths = jointMt.trie.getAllPaths(1);
  for (const path of paths) {
    if (path.length < 2) continue;
    if (path.every(id => (id & mask) !== 0)) {
      const strippedPath = path.map(id => id & ~mask);
      stripped._trie.insert(strippedPath);
      const srcNode = jointMt.trie.lookup(path);
      const dstNode = stripped._trie.lookup(strippedPath);
      if (srcNode && dstNode) {
        dstNode.visitCount = srcNode.visitCount;
        dstNode.predictionErrors = [...srcNode.predictionErrors];
        dstNode.meanPredictionError = srcNode.meanPredictionError;
      }
    }
  }

  if (jointMt.lastMetaState !== null && (jointMt.lastMetaState & mask) !== 0) {
    stripped._lastMetaState = jointMt.lastMetaState & ~mask;
  }

  for (const [mid, counts] of jointMt._tokenRegistry) {
    if ((mid & mask) !== 0) {
      stripped._tokenRegistry.set(mid & ~mask, new Map(counts));
    }
  }

  return stripped;
}

function fuse(agent) {
  if (agent.constituentIds.size === 0) return [agent];

  const [id1, id2] = [...agent.constituentIds];

  const mt1 = _splitMetaTrie(agent.experience.metaTrie, 0x10000000);
  const mt2 = _splitMetaTrie(agent.experience.metaTrie, 0x20000000);

  const _buildFusedAgent = (id, metaTrie) => {
    const ancestorLeafs = agent.leafConstituentIds;
    const directConstituents = agent.constituentIds;

    const leafIds = new Set();
    for (const leaf of ancestorLeafs) {
      leafIds.add(leaf);
    }

    const exp = new ExperienceSpace({
      trie: agent.experience.trie.merge(new ExperienceTrie(agent.experience.trie.maxDepth)),
      metaTrie,
      selfToken: new SelfTokenState(),
      lexicon: agent.experience.lexicon,
      traceBuffer: new TraceBuffer(agent.experience.traceBuffer.maxlen),
      lastWorldStateId: agent.experience.lastWorldStateId,
    });

    for (const e of [...agent.experience.traceBuffer]) {
      exp.traceBuffer.append(e);
    }

    return new ConsciousAgent({
      agentId: id,
      experience: exp,
      generation: agent.generation,
      constituentIds: new Set(),
      leafConstituentIds: leafIds,
      cycleLevel: Math.max(0, agent.cycleLevel - 1),
    });
  };

  return [_buildFusedAgent(id1, mt1), _buildFusedAgent(id2, mt2)];
}

module.exports = { combine, trivialAgent, experienceSpaceDistance, fuse, _splitMetaTrie };

File: hoffman-agents-python/src/conscious_agent/combination/operator.py
from __future__ import annotations

import hashlib

from ..core import (
    ExperienceTrie,
    MetaTrie,
    SelfTokenState,
    ExperienceLexicon,
    TraceBuffer,
    TraceEvent,
)
from ..agent import ConsciousAgent, ExperienceSpace, WorldState


def combine(*agents: ConsciousAgent) -> ConsciousAgent:
    if not agents:
        return trivial_agent()
    if len(agents) == 1:
        return agents[0]
    if len(agents) == 2:
        return _binary_combine(agents[0], agents[1])
    mid = len(agents) // 2
    return combine(combine(*agents[:mid]), combine(*agents[mid:]))


def _binary_combine(agent1: ConsciousAgent, agent2: ConsciousAgent) -> ConsciousAgent:
    if agent1.agent_id == "CA_0":
        return agent2
    if agent2.agent_id == "CA_0":
        return agent1

    combined_id = _hash_agent_ids(agent1.agent_id, agent2.agent_id)

    merged_trie = agent1.experience.trie.merge(agent2.experience.trie)

    leaf_ids_1 = agent1.leaf_constituent_ids or frozenset([agent1.agent_id])
    leaf_ids_2 = agent2.leaf_constituent_ids or frozenset([agent2.agent_id])
    leaf_constituent_ids = leaf_ids_1 | leaf_ids_2
    constituent_ids = frozenset({agent1.agent_id, agent2.agent_id})

    joint_mt = _build_joint_meta_trie(
        agent1.experience.meta_trie,
        agent2.experience.meta_trie,
    )

    combined_self = _combine_attractors(
        agent1.experience.self_token,
        agent2.experience.self_token,
    )

    merged_lexicon = _merge_lexicons(
        agent1.experience.lexicon,
        agent2.experience.lexicon,
    )

    init_state: int | None = None
    trace_buf = TraceBuffer(maxlen=50)
    if agent1.experience.last_world_state_id is not None or agent2.experience.last_world_state_id is not None:
        init_state = _canonical_hash(
            agent1.experience.last_world_state_id or -1,
            agent2.experience.last_world_state_id or -1,
        )
        trace_buf.append(TraceEvent(
            from_state=-1, to_state=init_state, timestamp=0,
            prediction=-1, prediction_correct=False, prediction_error=0.5,
        ))

    exp = ExperienceSpace(
        trie=merged_trie,
        meta_trie=joint_mt,
        self_token=combined_self,
        lexicon=merged_lexicon,
        trace_buffer=trace_buf,
        last_world_state_id=init_state,
    )

    return ConsciousAgent(
        agent_id=combined_id,
        experience=exp,
        generation=max(agent1.generation, agent2.generation),
        constituent_ids=constituent_ids,
        leaf_constituent_ids=leaf_constituent_ids,
        cycle_level=max(agent1.cycle_level, agent2.cycle_level) + 1,
    )


def trivial_agent() -> ConsciousAgent:
    return ConsciousAgent(agent_id="CA_0")


def _hash_agent_ids(id1: str, id2: str) -> str:
    data = f"{id1}->{id2}"
    h = hashlib.sha256(data.encode()).hexdigest()[:12]
    return f"CA_{h}"


def _canonical_hash(s1: int, s2: int) -> int:
    sorted_ids = sorted([s1, s2])
    data = f"ws:{sorted_ids[0]}:{sorted_ids[1]}"
    hash_bytes = hashlib.sha256(data.encode()).digest()
    return int.from_bytes(hash_bytes[:8], "big")


def _build_joint_meta_trie(mt1: MetaTrie, mt2: MetaTrie) -> MetaTrie:
    joint = MetaTrie(
        snapshot_window=max(mt1._snapshot_window, mt2._snapshot_window),
        max_depth=max(mt1.trie.max_depth, mt2.trie.max_depth),
    )

    for mid, snap in mt1._registry.items():
        joint._registry[mid | 0x10000000] = snap
    for mid, snap in mt2._registry.items():
        joint._registry[mid | 0x20000000] = snap

    if hasattr(mt1, "_token_registry"):
        for mid, counts in mt1._token_registry.items():
            joint._token_registry[mid | 0x10000000] = dict(counts)
    if hasattr(mt2, "_token_registry"):
        for mid, counts in mt2._token_registry.items():
            joint._token_registry[mid | 0x20000000] = dict(counts)

    return joint


def _combine_attractors(st1: SelfTokenState, st2: SelfTokenState) -> SelfTokenState:
    combined = SelfTokenState(
        token=st1.token,
        lock_threshold=min(st1.lock_threshold, st2.lock_threshold),
        lock_consecutive_required=max(st1.lock_consecutive_required, st2.lock_consecutive_required),
        stationary_prob=(st1.stationary_prob + st2.stationary_prob) / 2,
        locked=st1.locked and st2.locked,
    )
    if st1.locked and st2.locked:
        combined.referent_meta_state_id = st1.referent_meta_state_id
        combined.lock_generation = max(st1.lock_generation or 0, st2.lock_generation or 0)
    return combined


def _merge_lexicons(lex1: ExperienceLexicon, lex2: ExperienceLexicon) -> ExperienceLexicon:
    merged = ExperienceLexicon(
        embedding_dim=max(lex1.embedding_dim, lex2.embedding_dim),
        association_threshold=min(lex1._association_threshold, lex2._association_threshold),
    )
    for entry in list(lex1._entries.values()) + list(lex2._entries.values()):
        merged.bind(
            label=entry.label,
            output_token=entry.output_token,
            trace_signature=entry.trace_signature,
            prediction_error_peak=entry.prediction_error_peak,
            source=entry.labeling_source,
            generation=entry.generation_bound,
            step=entry.step_bound,
        )
        e = merged.lookup_by_label(entry.label)
        if e is not None:
            e.integration_depth = entry.integration_depth
            e.encounter_count = entry.encounter_count
    return merged


def experience_space_distance(exp1: ExperienceSpace, exp2: ExperienceSpace) -> float:
    p1 = set(tuple(p) for p in exp1.trie.get_all_paths(min_visits=0))
    p2 = set(tuple(p) for p in exp2.trie.get_all_paths(min_visits=0))
    union = p1 | p2
    if not union:
        return 0.0
    trie_dist = 1.0 - len(p1 & p2) / len(union)

    l1 = set(exp1.lexicon._entries.keys())
    l2 = set(exp2.lexicon._entries.keys())
    union_l = l1 | l2
    lexicon_dist = 0.0 if not union_l else 1.0 - len(l1 & l2) / len(union_l)

    return trie_dist * 0.6 + lexicon_dist * 0.4
